### PR TITLE
feat: Use journal LSNs for absolute replication offsets

### PR DESCRIPTION
**Pull request by royjacobson**
This does:

1. Switch dragonfly replication to use journal LSNs as replication offsets.
2. Implement changes to serialization of journal entries so both replica and master do the counting consistently.

### CODE REVIEW

File: src/server/dflycmd.cc [hunk 1]
```diff
@@ -354,23 +354,14 @@ void DflyCmd::Expire(CmdArgList args, ConnectionContext* cntx) {
...
 void DflyCmd::ReplicaOffset(CmdArgList args, ConnectionContext* cntx) {
   RedisReplyBuilder* rb = static_cast<RedisReplyBuilder*>(cntx->reply_builder());
-  string_view sync_id_str = ArgS(args, 1);
 
-  VLOG(1) << "Got DFLY REPLICAOFFSET " << sync_id_str;
-  auto [sync_id, replica_ptr] = GetReplicaInfoOrReply(sync_id_str, rb);
-  if (!sync_id)
-    return;
+  rb->StartArray(shard_set->size());
+  std::vector<LSN> LSNs(shard_set->size());
```
Comment (adiholden) [inline src/server/dflycmd.cc:359]: LSNs rename to lsns

File: src/server/dflycmd.cc [hunk 2]
```diff
@@ -354,23 +354,14 @@ void DflyCmd::Expire(CmdArgList args, ConnectionContext* cntx) {
...
-  unique_lock lk(replica_ptr->mu);
-  rb->StartArray(replica_ptr->flows.size());
-  for (size_t flow_id = 0; flow_id < replica_ptr->flows.size(); ++flow_id) {
-    JournalStreamer* streamer = replica_ptr->flows[flow_id].streamer.get();
-    if (streamer) {
-      rb->SendLong(streamer->GetRecordCount());
-    } else {
-      rb->SendLong(0);
-    }
+  for (size_t flow_id = 0; flow_id < shard_set->size(); ++flow_id) {
```
Comment (adiholden) [inline src/server/dflycmd.cc:363]: I would rename flow_id to shard_id, the lsn number is per shard and not per replication flow

File: src/server/dflycmd.cc [hunk 3]
```diff
@@ -354,23 +354,14 @@ void DflyCmd::Expire(CmdArgList args, ConnectionContext* cntx) {
...
-  string_view sync_id_str = ArgS(args, 1);
 
-  VLOG(1) << "Got DFLY REPLICAOFFSET " << sync_id_str;
-  auto [sync_id, replica_ptr] = GetReplicaInfoOrReply(sync_id_str, rb);
-  if (!sync_id)
-    return;
+  rb->StartArray(shard_set->size());
+  std::vector<LSN> LSNs(shard_set->size());
+  shard_set->RunBriefInParallel(
+      [&](EngineShard* shard) { LSNs[shard->shard_id()] = sf_->journal()->GetLsn(); });
```
Comment (adiholden) [inline src/server/dflycmd.cc:361]: journal() can return nullptr

File: src/server/rdb_load.cc [hunk 4]
```diff
@@ -2049,8 +2051,11 @@ error_code RdbLoaderBase::HandleJournalBlob(Service* service) {
   while (done < num_entries) {
     journal::ParsedEntry entry{};
     SET_OR_RETURN(journal_reader_.ReadEntry(), entry);
+    if (entry.opcode != journal::Op::EXEC)
```
Comment (adiholden) [inline src/server/rdb_load.cc:2054]: redundant line

File: tests/dragonfly/replication_test.py [hunk 5]
```diff
@@ -83,7 +83,7 @@ async def run_replication(c_replica):
 
 async def check_replica_finished_exec(c_replica, c_master):
     syncid, r_offset = await c_replica.execute_command("DEBUG REPLICA OFFSET")
-    command = "DFLY REPLICAOFFSET " + syncid.decode()
+    command = "DFLY REPLICAOFFSET"
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:86]: m_offset = await c_master.execute_command("DFLY REPLICAOFFSET")

File: src/server/rdb_save.h [hunk 6]
```diff
@@ -155,7 +155,7 @@ class RdbSerializer {
   std::error_code WriteJournalEntry(const journal::Entry& entry);
 
   // Send FULL_SYNC_CUT opcode to notify that all static data was sent.
-  std::error_code SendFullSyncCut();
+  std::error_code SendFullSyncCut(uint64_t final_version);
```
Comment (dranikpg) [inline src/server/rdb_save.h:158]: Just by its name its hard to guess what `final_version` really means

File: src/server/rdb_save.cc [hunk 7]
```diff
@@ -666,14 +666,16 @@ error_code RdbSerializer::SaveStreamConsumers(streamCG* cg) {
   return error_code{};
 }
 
-error_code RdbSerializer::SendFullSyncCut() {
+error_code RdbSerializer::SendFullSyncCut(uint64_t final_version) {
   RETURN_ON_ERR(WriteOpcode(RDB_OPCODE_FULLSYNC_END));
 
+  VLOG(1) << "final_version=" << final_version;
   // RDB_OPCODE_FULLSYNC_END followed by 8 bytes of 0.
```
Comment (adiholden) [inline src/server/rdb_save.cc:673]: this comment can be deleted

File: src/server/rdb_save.cc [hunk 8]
```diff
@@ -666,14 +666,16 @@ error_code RdbSerializer::SaveStreamConsumers(streamCG* cg) {
...
   // The reason for this is that some opcodes require to have at least 8 bytes of data
   // in the read buffer when consuming the rdb data, and since RDB_OPCODE_FULLSYNC_END is one of
   // the last opcodes sent to replica, we respect this requirement by sending a blob of 8 bytes.
-  uint8_t buf[8] = {0};
+  uint8_t buf[sizeof(uint64_t)] = {0};
```
Comment (adiholden) [inline src/server/rdb_save.cc:677]: no so sure I understand this line. You want to write uint64_t final_version
why do you do uint8_t buf[sizeof(uint64_t)] = {0};

File: src/server/rdb_load.cc [hunk 9]
```diff
@@ -2049,8 +2051,11 @@ error_code RdbLoaderBase::HandleJournalBlob(Service* service) {
...
     ex.Execute(entry.dbid, entry.cmd);
+    VLOG(1) << "Reading item [" << journal_offset_ << "]: " << entry.ToString();
     done++;
+    journal_offset_++;
```
Comment (adiholden) [inline src/server/rdb_load.cc:2058]: you override this in RDB_OPCODE_FULLSYNC_END
why do you need it here?

File: src/server/rdb_load.cc (hunk 4 above)
Comment (royjacobson) [reply src/server/rdb_load.cc:2054]: it's not redundant, just messed up formatting :sweat_smile:

File: src/server/rdb_save.cc (hunk 8 above)
Comment (royjacobson) [reply src/server/rdb_save.cc:677]: Hmm, I probably should fix it to write it through the endianity functions instead anyway. But what should I do instead?

File: src/server/rdb_load.cc (hunk 9 above)
Comment (royjacobson) [reply src/server/rdb_load.cc:2058]: We send journal updates in the full sync even after `RDB_OPCODE_FULLSYNC_END`. And I have to keep track of that :/

File: src/server/snapshot.cc [hunk 10]
```diff
@@ -259,14 +270,15 @@ void SliceSnapshot::OnDbChange(DbIndex db_index, const DbSlice::ChangeReq& req)
 // OnJournalEntry registers for changes in journal, the journal change function signature is
 // (const journal::Entry& entry, bool await) In snapshot flow we dont use the await argument.
 void SliceSnapshot::OnJournalEntry(const journal::Entry& entry, bool unused_await_arg) {
-  // We ignore non payload entries like EXEC because we have no transactional ordering during
-  // LOAD phase on replica.
-  if (!entry.HasPayload()) {
+  // We ignore non payload entries because we have no transactional ordering during
+  // LOAD phase on replica, but we have to count EXEC because they affect the journal
+  // counters that have to stay synchronized in replication.
```
[the remark this answers is not in the document]
Comment (royjacobson) [reply src/server/snapshot.cc:275]: We can receive updates through the full sync even after a full sync cut. We could maybe delay the sync cut, or try to send the LSN through some other mechanism, but the way this currently works I don't see a way around it :/

File: src/server/rdb_load.cc (hunk 9 above)
Comment (adiholden) [reply src/server/rdb_load.cc:2058]: Is it possible to send the the lsn in reply for "DFLY FLOW"  sent from StartFullSyncFlow in replica?

Comment (adiholden) [reply src/server/rdb_load.cc]: are you going to fix the flow to send it in the DFLY STARTSTABLE?

File: src/server/replica.cc [hunk 11]
```diff
@@ -888,6 +888,7 @@ void Replica::FullSyncDflyFb(string eof_token, BlockingCounter bc, Context* cntx
     leftover_buf_.reset();
   }
 
+  this->journal_rec_executed_.store(loader.journal_offset());
```
Comment (adiholden) [inline src/server/replica.cc:891]: Please update the comment in replica.h for this member

**Review by adiholden — APPROVED** [review]

**Pull request merged**